Annotate endpoints and verbs with Javadoc comments

brainwane opened the pull request:
As a stopgap measure to give us some automated API documentation,
I've annotated the @RequestMapping definitions with custom Javadoc
tags `@endpoint` and `@verb`. To generate the documentation in a new
"doc-endpoints" directory, invoke Javadoc from the psm-app
directory:

`$ javadoc -d doc-endpoints/cms-web/ -sourcepath cms-web/src/main/java/ -subpackages . -tag endpoint:a:"Endpoint: " -tag verb:a:"HTTP Verbs Allowed: "`

Related to #30.

cecilia-donnelly commented:
@brainwane thanks for this!  It looks great.  Can you add the `doc-endpoints` dir to .gitignore?

brainwane commented:
Done! Thanks @cecilia-donnelly.

jasonaowen commented on psm-app/cms-web/src/main/java/gov/medicaid/controllers/EnrollmentPageFlowController.java:
```diff
@@ -112,6 +112,7 @@
  * 
  * @author TCSASSEMBLER
  * @version 1.0
+ * @endpoint "/provider/enrollment/*"
```
Including these quotes results in quotes in the rendered Javadoc - is that what you want?

brainwane replied:
I'm good with the quotes but thank you for checking.

jasonaowen commented on psm-app/cms-web/src/main/java/gov/medicaid/controllers/EnrollmentPageFlowController.java:
```diff
@@ -214,6 +215,8 @@ protected void init() {
      *             for read/write errors
      * @throws PortalServiceException
      *             for any other errors
+     * @endpoint "/attachment"
```
I believe this method's endpoint (`/attachment`) is relative to the class's endpoint (`/provider/enrollment`), so the overall endpoint is `/provider/enrollment/attachment`. Do you want to include that in the tag? Or do you feel like it's clear in context?

brainwane replied:
Cecilia and I just spoke in IRC and decided I should add the relevant roots to make sure that we display full endpoints to readers. So I'll modify the relevant values in tags like this. Thanks for catching this.

cecilia-donnelly commented on psm-app/cms-web/src/main/java/gov/medicaid/controllers/EnrollmentPageFlowController.java:
```diff
@@ -475,6 +492,8 @@ public ModelAndView selectProviderType(@ModelAttribute("enrollment") EnrollmentT
      * @return the previous page, or the same page if there were errors
      * @throws PortalServiceException
      *             for any errors encountered
+     * @endpoin "/steps/prev"
```
This one looks like it got interrupted -- needs a "t" and a `/provider/enrollment`.

brainwane replied:
Thanks! Fixed.

cecilia-donnelly commented on psm-app/cms-web/src/main/java/gov/medicaid/controllers/ForgotPasswordController.java:
```diff
@@ -77,6 +78,7 @@ protected void init() {
      *
      * @return the form entry.
      * @throws PortalServiceException for any errors encountered
+     * @verb GET
```
@brainwane and @jasonaowen, you talked and decided that these both inherit `/forgotpassword`, right?  So we don't need to specify?

brainwane replied:
I did not discuss this with Jason, and would welcome his opinion here!

jasonaowen replied:
That's the behavior I expect; I can go either way on the annotation.

cecilia-donnelly commented on psm-app/cms-web/src/main/java/gov/medicaid/controllers/admin/EnrollmentController.java:
```diff
@@ -326,6 +333,7 @@ public ModelAndView getHelpItem(@RequestParam("helpItemId") long id) throws Port
      *            the ticket id
      * @return the model and view for
      * @throws PortalServiceException
+     * @endpoint "/agent/enrollment/rejectTicket"
```
Do we know why this one and the next two don't specify a method?

brainwane replied:
Discusssed https://github.com/OpenTechStrategies/psm/pull/221#discussion_r124398876

cecilia-donnelly commented on psm-app/cms-web/src/main/java/gov/medicaid/controllers/admin/EnrollmentController.java:
```diff
@@ -478,6 +488,8 @@ public ModelAndView viewScreeningLog(@RequestParam("id") long id, @RequestParam(
      * @throws PortalServiceException
      *             - If there are any errors in the action
+     * @endpoint "/agent/enrollment/exportBatch"
+     * @endpoint "/provider/search/exportBatch"
```
Do we know why there are two endpoints?

jasonaowen replied:
I would guess so that the same method can be used by both providers and service agents. Doing a `git grep exportBatch` shows that only the former has an explicit link from a JSP, so it looks like the latter is unused.

jasonaowen commented on psm-app/cms-web/src/main/java/gov/medicaid/controllers/admin/EnrollmentController.java:
```diff
@@ -214,6 +214,7 @@ public void initBinder(WebDataBinder binder) {
      *             - If there are any errors in the action
+     * @endpoint "/agent/enrollment/"
      */
     @RequestMapping("/agent/enrollment/")
```
Hm, this doesn't have a `method`! From my reading of [the Spring reference docs](http://docs.spring.io/spring/docs/3.0.5.RELEASE/spring-framework-reference/html/mvc.html#mvc-ann-requestmapping), it seems that when no verb is specified the default is to route all requests, regardless of verb, to this method. Should we indicate that with a Javadoc tag? Several of the other methods in this controller also omit a `method`, so if so we should update all of them.

brainwane replied:
In this controller and several other files affected by this PR, where there is no `method` explicitly listed, I've left out the `@verb` tag. Right now I think it is better to leave out `@verb`, and avoid setting an expectation that all verbs will hit this method, unless and until we test the application to check that it's true.

jasonaowen commented on psm-app/cms-web/src/main/java/gov/medicaid/controllers/admin/SystemAdminUserController.java:
```diff
@@ -91,6 +92,7 @@ protected void init() {
      * @throws PortalServiceException - If there are any errors in the action
+     * @endpoint "/system/user/list"
      */
     @RequestMapping("/list")
```
This controller has several missing `@verb`s, as in the other controller.

brainwane replied:
As discussed in https://github.com/OpenTechStrategies/psm/pull/221#discussion_r124398876 .

jasonaowen commented on psm-app/cms-web/src/main/java/gov/medicaid/controllers/admin/SystemAdminUserSearchController.java:
```diff
@@ -68,6 +69,7 @@ public SystemAdminUserSearchController() {
      * @throws PortalServiceException - If there are any errors in the action
+     * @endpoint "/system/search/list"
      */
     @RequestMapping("/list")
```
Missing `@verb`s.

brainwane replied:
As discussed in https://github.com/OpenTechStrategies/psm/pull/221#discussion_r124398876 .

jasonaowen commented:
Your PR description has [the magic words that will close an issue upon merge](https://help.github.com/articles/closing-issues-via-commit-messages/) - if you want the issue to remain open, you should change the last line of the description.

brainwane commented:
Thanks for the suggestions, @jasonaowen - I've fixed the commit message and I've added document generation instructions to the `INSTALL.md` file. When I make a PR to do the Travis autogeneration of docs I'll put a line in `README.md` pointing people to our GitHub Pages site.

jasonaowen commented:
@brainwane you'll also need to change the PR description - right now, the issue will be closed when this PR is merged. This is reflected in the GitHub UI in the issue timeline on the line showing the reference to the issue from this PR by the little `(i)`.

brainwane commented:
Thanks @jasonaowen -- addressed.

brainwane commented on psm-app/cms-web/src/main/java/gov/medicaid/controllers/SelfRegistrationController.java:
```diff
@@ -82,6 +83,8 @@ protected void init() {
      * Loads the registration form.
      *
      * @return the view and model for the registration page.
+     * @endpoint "/accounts/new"
```
So this seems like an irregularity in the code; what do you think we ought to do for the annotation?

jasonaowen replied:
Since you're marking full endpoints, I think you're good. It's just a part of the application I don't understand yet, so I was thinking aloud a bit and hoping you might have read something about this while you were researching Spring stuff. I don't think we need to change anything!
